### PR TITLE
Fix fransport handshake starting before tls handshake completes

### LINKED ISSUE #77999: Transport handshake starts before TLS handshake completes
When opening a connection to a remote node we establish a collection of TCP channels and once they're all open we send an `internal:tcp/handshake` to the remote node with a 30-second timeout. However today we don't wait for the TLS handshake to complete before considering a channel to be open. This means if the TLS handshake fails (e.g. it times out after 10s) then we report the `internal:tcp/handshake` as failing due to the channel closing, rather than reporting a more useful error about the TLS handshake failure. We should wait for the TLS handshake to complete before considering each channel to be open.

### PR DESCRIPTION
**Pull request by original-brownbear**
Implements waiting for the TLS handshake to complete before triggering the connected callback and thus the transport handshake.
Had to make copies of two transport test cases with slight but not trivial alterations to force a tls handshake before doing either a timeout or a connection reset. Will dry that up in a follow-up as it's a bit a noisy refactoring.

closes #77999

### CODE REVIEW

**Comment by original-brownbear** [comment]
@elasticmachine update branch

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/core/security/transport/netty4/SecurityNetty4Transport.java [hunk 1]
```diff
@@ -217,8 +218,24 @@ public void connect(ChannelHandlerContext ctx, SocketAddress remoteAddress, Sock
...
+            final Future<?> handshakePromise = sslHandler.handshakeFuture();
+            connectPromise.addListener(result -> {
+                if (result.isSuccess() == false) {
+                    promise.tryFailure(result.cause());
+                } else {
+                    handshakePromise.addListener(handshakeResult -> {
+                        if (handshakeResult.isSuccess()) {
+                            promise.setSuccess();
+                        } else {
+                            promise.tryFailure(handshakeResult.cause());
```
Comment (original-brownbear) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/core/security/transport/netty4/SecurityNetty4Transport.java:233]: Went for `tryFailure` here out of paranoia, probably not necessary but didn't want to risk an uncaught exception here.

**Comment by original-brownbear** [comment]
@DaveCTurner mind taking a look here, I think this one's still somewhat valuable :)

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
Thanks David!

**Pull request merged**